cpp/src/search/contains_table.cu

```diff
@@ -1,148 +1,378 @@
+                                    size_type,
+                                    cuda::thread_scope_device,
+                                    rmm::mr::stream_allocator_adaptor<default_allocator<char>>>;
+
+/**
+ * @brief An adapter functor to support strong index types for row hasher that must be operating on
+ * `cudf::size_type`.
+ */
+template <typename Hasher>
+struct strong_index_hasher_adapter {
+  strong_index_hasher_adapter(Hasher const& hasher) : _hasher{hasher} {}
+
+  template <typename T>
+  __device__ inline auto operator()(T const idx) const noexcept
+  {
+    return _hasher(static_cast<size_type>(idx));
+  }
+
+ private:
+  Hasher _hasher;
+};
+
+/**
+ * @brief An adapter functor to support strong index type for table row comparator that must be
+ * operating on `cudf::size_type`.
+ */
+template <typename Comparator>
+struct strong_index_comparator_adapter {
+  strong_index_comparator_adapter(Comparator const& comparator) : _comparator{comparator} {}
+
+  template <typename T>
+  __device__ inline auto operator()(T const lhs_index, T const rhs_index) const noexcept
+  {
+    return _comparator(static_cast<size_type>(lhs_index), static_cast<size_type>(rhs_index));
+  }
+
+  __device__ inline auto operator()(lhs_index_type const lhs_index,
+                                    rhs_index_type const rhs_index) const noexcept
+  {
+    return _comparator(static_cast<size_type>(lhs_index), static_cast<size_type>(rhs_index));
+  }
+
+  // This overload enforces symmetry for the two table comparator that doesn't support strong index
+  // types. When the indices are provided in wrong order, this overload switches them back to the
+  // right one.
+  __device__ inline auto operator()(rhs_index_type const rhs_index,
+                                    lhs_index_type const lhs_index) const noexcept
+  {
+    return _comparator(static_cast<size_type>(lhs_index), static_cast<size_type>(rhs_index));
+  }
+
+ private:
+  Comparator const _comparator;
+};
+
+/**
+ * @brief Invoke an `operator()` template with a row equality comparator based on the specified
+ * `compare_nans` parameter.
+ *
+ * @param compare_nans The flag to specify whether NaNs should be compared equal or not
+ * @param func The input functor to invoke
+ */
+template <typename Func>
+void dispatch_nan_comparator(nan_equality compare_nans, Func&& func)
+{
+  if (compare_nans == nan_equality::ALL_EQUAL) {
+    using nan_equal_comparator =
+      cudf::experimental::row::equality::nan_equal_physical_equality_comparator;
+    func(nan_equal_comparator{});
+  } else {
+    using nan_unequal_comparator = cudf::experimental::row::equality::physical_equality_comparator;
+    func(nan_unequal_comparator{});
+  }
+}
+
+/**
+ * @brief Check if the input table has any lists column.
+ *
+ * @param input The input table
+ * @return A boolean indicating if the input table has any lists column
+ */
+inline bool has_nested_lists(table_view const& input)

```

Comment:
This is [duplicated code from `cudf::structs::detail::is_or_has_nested_lists`](https://github.com/rapidsai/cudf/blob/ebcea0fa1edba776a1631aac4b7e50b8127a3e4f/cpp/src/structs/utilities.cpp#L81-L86). Perhaps we can move that out of the structs namespace to a general utilities file and reuse it here. I'd be okay with doing this in a follow-up PR since it may touch a few files, but would like to see it done eventually.